### PR TITLE
hubble: Add support for SockLB tracing

### PR DESCRIPTION
**Pull request by gandro**
This PR adds a Hubble parser for `TraceSockNotify` events. Those
events are emitted whenever SockLB is performed, see cilium/cilium#20492
for details about the datapath tracing.

This parser is similar to the L3/L4 event parser. But because the events
are traced on the socket level, we do not know the source endpoint id,
source endpoint IP address or source TCP/UDP port of the event. We only
know the source cgroup ID. From the cgroup, we can derive which local
pod emitted the event (via `GetPodMetadataForContainer`). From the pod
IP we can then perform an IP-based lookup in either the local endpoint
list or the IPCache to determine the source endpoint metadata (endpoint
id, numeric identity, labels, etc). The lookup for destination IP is
almost identical to the threefour parser, with the exception that we do
not have any datapath identity available for fallback.

The resulting flow looks similar to regular `FlowType_L3_L4`, with a
couple of key differences:

 - The cgroup id of the container is available
 - Information if the event is pre or post translation point is
   available
 - The socket cookie is available
 - The TCP/UDP source port is _not_ known
 - TCP flags are _not_ known
 - Ethernet headers are _not_ known

Unfortunately, sock events can also be performed by non-pod entities
(e.g. the host cgroup). In that case, we cannot perform a mapping to any
local endpoint and therefore the source IP and endpoint data are left
empty. The cgroup id may also not be available on older kernels or
systems with the systemd cgroup driver (though this will likely be fixed
in a follow-up PR). This fact that the source IP may be empty/unknown is
also key difference to the flows which have been produced by Hubble so
far.

---

Here's an example of a pod performing a DNS request and response. The first three events are the DNS request: We first see two the SockLB events (pre and post translation, meaning before and after service backend selection):

image

And here are the events for the response: We see the response packet arriving at the endpoint, but before it is delivered over the socket, reverse NAT being applied, i.e. the pod IP is once again translated to the service IP (which is where the userspace sent the packet to):

image

### CODE REVIEW

File: pkg/hubble/parser/sock/parser.go
Comment (aditighag) [inline pkg/hubble/parser/sock/parser.go]: Can we remove the hard-coded forwarded verdict? This policy related field isn't relevant for socket-lb tracing events.

File: pkg/hubble/parser/sock/parser.go
Comment (aditighag) [inline pkg/hubble/parser/sock/parser.go:109]: What's `SourceService` field for?

File: pkg/hubble/parser/sock/parser.go
Comment (aditighag) [inline pkg/hubble/parser/sock/parser.go:130]: When would `p.cgroupGetter` be `nil`? We should fail more assertively instead of having a check for every trace event if it's `nil`.

File: pkg/hubble/parser/sock/parser.go
Comment (aditighag) [inline pkg/hubble/parser/sock/parser.go:131]: Hmm... We should prune calls from host processes (essentially non-pods). Host cgroup id will usually be `1`, except for kind like nested environments. Although there can also be events initiated from non-pod entities like `kubelet` which can have its own cgroup id. 

Maybe datapath can assist here by indicating whether events are initiated from the host network namespace? We already have logic baked into the datapath that identified "host" network namespace on nested environments like kind.

File: pkg/hubble/parser/sock/parser.go
Comment (aditighag) [inline pkg/hubble/parser/sock/parser.go:201]: Fancy!

File: pkg/hubble/parser/sock/parser.go
Comment (aditighag) [inline pkg/hubble/parser/sock/parser.go:245]: Summary is printing only the L4 protocol. Is this code complete?

**Review by aditighag — CHANGES_REQUESTED** [review]
🚀 I only looked at the commit "hubble: Add support for SockLB tracing" that pertains to parsing socket-lb trace events. The parsing logic looks good. Let's follow up on pruning events from non-pod entities. 

Can we print cgroup id like metadata in the verbose json output? Will you be adding json summary in a follow-up?

File: pkg/hubble/parser/sock/parser.go
Comment (gandro) [reply pkg/hubble/parser/sock/parser.go]: I'm not sure if the trace sock events are fundamentally different from e.g. trace events emitted by FROM_LXC. Those also are also emitted way before policy is applied and contain the same verdict. 

In general, I agree with your sentiment, but in the context of the other Hubble events we have, I do think FORWARDED makes somewhat sense here. We could add a TRANSLATED verdict, maybe?

File: pkg/hubble/parser/sock/parser.go
Comment (gandro) [reply pkg/hubble/parser/sock/parser.go:109]: Note that we swap source and destination for rev nat events. This therefore can be the service name of the "DstIP". 
Check the third flow of the second screenshot above, the `default/kube-dns:53` service name is what's being populated here.

File: pkg/hubble/parser/sock/parser.go
Comment (gandro) [reply pkg/hubble/parser/sock/parser.go:130]: It's only nil in unit testing. Other Hubble code does the same. The reason being that we've had bug reports of OSS Fuzz: https://github.com/cilium/cilium/pull/20446

Arguably a bug in their framework, but not requiring this to be set simplifies fuzzing nonetheless.

File: pkg/hubble/parser/sock/parser.go
Comment (gandro) [reply pkg/hubble/parser/sock/parser.go:131]: Hm. Hubble so far does not do any filtering of events by default, it forwards everything that it receives and asks the user to filter out irrelevant events.

I would therefore also prefer to filter this in the datapath. If that's not possible and we need to teach Hubble to drop events here, then we need to discuss how to best do this. But I do prefer for `hubble observe` and `cilium monitor` to always display the same events (which is the case right now).

File: pkg/hubble/parser/sock/parser.go
Comment (gandro) [reply pkg/hubble/parser/sock/parser.go:245]: Yes. The summary refers to the string in parentheses in the following screenshot (e.g. `ICMPv6 RouterSolicitation`)

image

Since we don't have more information on the actual contents of the packet for trace sock events (e.g. TCP flags), just displaying the protocol seems to be the most consistent with the existing output. But I'm open to suggestions.

**Comment by gandro** [comment]
> Can we print cgroup id like metadata in the verbose json output? Will you be adding json summary in a follow-up?

This already done. The JSON schema is derived from the protobuf schema, which is updated in this PR to contain the cgroup id. Once the schema is synced to the Hubble CLI (which we can only do once this PR is merged), `hubble observe -o jsonpb` will print it too.

File: pkg/hubble/parser/sock/parser.go
Comment (gandro) [reply pkg/hubble/parser/sock/parser.go]: Discussed with the Hubble team, we'll add a new verdict kind for those events.

File: pkg/hubble/parser/sock/parser.go
Comment (gandro) [reply pkg/hubble/parser/sock/parser.go:131]: Discussed with the Hubble team: We'll filter out any events without a matching pod in Hubble, since those events don't provide much value to end users. There will be an agent option to disable this filtering for developers and troubleshooting (but the raw data should also be available via cilium monitor).

File: pkg/hubble/parser/common/labels.go
Comment (kaworu) [inline pkg/hubble/parser/common/labels.go]: nit: maybe move this branch as the first one in the for loop to reduce indentation, i.e. something like

```go
for _, label := range labels {
    if !strings.HasPrefix(label, cidrPrefix) {
        filteredLabels = append(filteredLabels, label)
        continue
    }
    // max cidr label logic here
}
```

File: pkg/hubble/parser/common/endpoint.go
Comment (kaworu) [inline pkg/hubble/parser/common/endpoint.go]: nit
```suggestion
		userspaceSecurityIdentity := identity.Uint32()
```

File: pkg/hubble/parser/common/labels.go
Comment (kaworu) [inline pkg/hubble/parser/common/labels.go]: nit: IIUC technically we don't need to remember the actual max `IPNet` here, only its mask size. Assuming `var max *net.IPNet` was replaced by `var maxSize int` earlier we could

```suggestion
			if currMask, _ := curr.Mask.Size(); currMask > maxSize {
				maxSize, maxStr = currMask, label
			}
```

This simplify the logic a bit and also avoid recomputing the max mask size.

File: pkg/hubble/parser/parser.go
Comment (kaworu) [inline pkg/hubble/parser/parser.go]: Seems to me that this comment isn't accurate anymore after this patch as we can get [a TraceSock event on connect](https://github.com/cilium/cilium/blob/7cc3e3a3e2f60aba31d4985cb986c98bfc4dae34/bpf/bpf_sock.c#L464) correct?

File: pkg/hubble/parser/sock/parser.go
Comment (kaworu) [inline pkg/hubble/parser/sock/parser.go]: At first I thought `decodeRevNat` returning a `*wrapperspb.BoolValue` was a left-over from a previous version of the patch, because `isRevNat` is never added to the decoded flow and this line is the only call site. We check for `isRevNat.GetValue()` a few lines below, but we only need a `true`/`false` signal at that point.

Looking at the screenshots of the Hubble CLI output, I noticed that we're missing the arrows for `TraceSock` flows (i.e. we see `<>`), so related question: would setting `IsReply` and `TrafficDirection` based on `isRevNat` make sense for `TraceSock` events?

File: pkg/hubble/parser/sock/parser.go
Comment (kaworu) [reply pkg/hubble/parser/sock/parser.go:245]: `Summary` is marked as deprecated, and `decodeSummary` only provides info that we already know through `L4`. IMHO this logic belongs to the client, but I don't feel too strongly about it.

**Review by kaworu — COMMENTED** [review]
Awesome work @gandro! I have a few comments, mostly cosmetic. Feel free to ignore nits and comments on code that has only moved (I only noticed after review).

File: pkg/hubble/parser/sock/parser.go
Comment (gandro) [reply pkg/hubble/parser/sock/parser.go:131]: Added a flag to filter them out (see last commit).

File: pkg/hubble/parser/sock/parser.go
Comment (gandro) [reply pkg/hubble/parser/sock/parser.go]: Added the `TRANSLATED` verdict

File: pkg/hubble/parser/sock/parser.go
Comment (gandro) [reply pkg/hubble/parser/sock/parser.go]: Yes, your suspicion is correct, an earlier version of the PR used the result of  `decodeRevNat` as the reply flag. I've changed it to a `bool` in the most recent version.

Unfortunately, trying to derive the traffic direction / reply state from the observation point does not work: `IsReply` cannot be meaningfully populated for TraceSock events because their occur on the socket granularity level, not a packet level. 

The screenshot in the PR description only shows the traces from the node where `pod-to-a` is hosted, where rev nat is indeed only used for the reply. But if we include also the events emitted by the kube-dns host, then the picture looks slightly different, we also see rev nat being used for the non-reply packet:

image

The screenshot above is using `xwing`, but it's a single UDP request packet being sent from xwing to kubedns, with a single reply packet being sent back. Please note the timetsamps - as you can see, `kind-worker2` (where coredns is hostd) emits a `pre-xlate-rev` for the initial incoming packet (there is no post event, as there is no translation). The reply packet (the last 4 events) happens half a second later, and does not emit any SockLB event, probably because it's sent back on the same "connected UDP socket".

**Comment by gandro** [comment]
Thanks for the feedback @aditighag and @kaworu - I think I addressed or commented on all of it.

File: pkg/hubble/parser/sock/parser.go
Comment (gandro) [reply pkg/hubble/parser/sock/parser.go]: Another way to formulate this: I think (but I might be mistaken, cc @aditighag) that `pre/post-xlate-rev` is only executed on `recvmsg` and `getpeername`, both system calls don't really imply anything of overall the "traffic direction" (e.g. if it was an ingress or egress connection in policy lingo).

`pre/post-xlate-fwd` on the other hand is executed for `connect` and `sendmsg`. `connect` is typically used to initiate an egress connection, but the same cannot necessary be said for `sendmsg`.

**Review by jibi — COMMENTED** [review]
Agent changes look good, but I think we are missing the related changes for the Helm charts

**Comment by gandro** [comment]
> 

Good point. I originally though to omit it (since I expect the newly added option only to be relevant to devs), but it might be still worth adding it nonetheless, since `extraConfig` is rarely known. Will add a commit adding the Helm value.

**Review by jibi — APPROVED** [review]

**Review by michi-covalent — APPROVED** [review]

File: pkg/hubble/parser/sock/parser.go
Comment (kaworu) [reply pkg/hubble/parser/sock/parser.go]: @gandro ah that make sense, thanks for the detailed explanation :+1: one more question about the example:

> The reply packet (the last 4 events) happens half a second later, and does not emit any SockLB event, probably because it's sent back on the same "connected UDP socket".

To make sure I understand correctly: the reply packet does not emit any SockLB event on `kind-worker2` (i.e. `coredns` host), but it does in `kind-worker` (i.e. `xwing` host) correct? The screenshot's last two events are a couple of `pre/post-xlate-rev` analogous to the request packet `pre/post-xlate-fwd` events (which make sense to me), hence the question.

File: pkg/hubble/observer/local_observer.go
Comment (kaworu) [inline pkg/hubble/observer/local_observer.go]: nit: could handle the invalid type case in the `switch` directly

```suggestion
			switch {
			// silently ignore those kind of errors
			case errors.Is(err, parserErrors.ErrUnknownEventType),
				errors.Is(err, parserErrors.ErrEventSkipped),
				// silently ignore perf ring buffer events with invalid types,
				// since they are not intended for us (e.g. MessageTypeRecCapture)
				parserErrors.IsErrInvalidType(err):
			default:
				s.log.WithError(err).WithField("event", monitorEvent).Debug("failed to decode payload")
```

**Review by kaworu — APPROVED** [review]
:100:

File: pkg/hubble/observer/local_observer.go
Comment (gandro) [reply pkg/hubble/observer/local_observer.go]: Good suggestion, applied!

File: pkg/hubble/parser/sock/parser.go
Comment (gandro) [reply pkg/hubble/parser/sock/parser.go]: That's correct, yes. My guess here is that `coredns` is using Golang's `net.Conn` to send the reply packet, meaning it's using the `write` system call on a "connected UDP socket". Thus, none of the above cgroup hooks is hit. `dig` (used inside `xwing`) on the other hand might be using `readmsg`/`sendmsg` syscalls, thus causing a Hubble event for each syscall.

Maybe @aditighag has deeper insight here.

Comment (gandro) [reply pkg/hubble/parser/sock/parser.go]: On second thought, connected UDP sockets are only useful for clients. Checking a basic UDP echo server, it seems Go is using `recvfrom` and `sendto`, so not sure why the `sendto` does not cause a Hubble event (it should for unconnected udp sockets, according to https://manpages.debian.org/experimental/bpftool/bpftool-cgroup.8.en.html)

**Comment by gandro** [comment]
- [k8s-1.24-kernel-4.19](https://jenkins.cilium.io/job/Cilium-PR-K8s-1.24-kernel-4.19/743/) hit known flake https://github.com/cilium/cilium/issues/21979
- [CI Multicluster](https://github.com/cilium/cilium/actions/runs/3379802667/jobs/5611843502) seems to have hit a variant of https://github.com/cilium/cilium/issues/21655 (in any case, it's a datapath flake which this PR does not touch).

File: pkg/hubble/parser/sock/parser.go
Comment (aditighag) [reply pkg/hubble/parser/sock/parser.go]: The recvmsg hook is executed only for regular UDP. TCP (and connected UDP) sockets are connected, so TCP clients send and receive data using `send`/`recv` socket calls.
@gandro Your observation is correct. Golang uses a connected UDP client by default. Regarding, `sendto` for connected UDP, here is a snippet from the man page - 

`If sendto() is used on a connection-mode (SOCK_STREAM, SOCK_SEQPACKET) socket, the arguments dest_addr and addrlen are ignored (and the error EISCONN may be returned when they are not NULL and 0)`

Comment (aditighag) [reply pkg/hubble/parser/sock/parser.go]: >Another way to formulate this: I think (but I might be mistaken, cc @aditighag) that pre/post-xlate-rev is only executed on recvmsg and getpeername, both system calls don't really imply anything of overall the "traffic direction" (e.g. if it was an ingress or egress connection in policy lingo).

When podA on nodeA talks to podB on nodeB, both fwd and rev translation happen on the nodeA (for UDP). For TCP and connected UDP, it would just be fwd translation.  

The {pre, post}-fwd/rev markers already convey the direction. We shouldn't try to interpret this with the same context as policy terminology (ingress/egress).

fwd: svc vip -> backend ip
rev: backend ip -> svc vip

Hope that makes sense.

File: pkg/hubble/parser/sock/parser.go
Comment (aditighag) [reply pkg/hubble/parser/sock/parser.go:130]: I expressed the same concern as Joe did on the PR you linked - https://github.com/cilium/cilium/pull/20446#discussion_r918669500. 
It doesn't look like we followed up doing `nil` checks in constructors that Tobias proposed, did we?

File: pkg/hubble/parser/sock/parser.go
Comment (aditighag) [inline pkg/hubble/parser/sock/parser.go]: We should set this only on the post_* notifications. This isn't applicable for pre_* events.

**Review by aditighag — CHANGES_REQUESTED** [review]
Commit "hubble: Add support for SockLB tracing" looks good to me. Nice work! 

One minor comment around setting the verdict field.

File: pkg/hubble/parser/sock/parser.go
Comment (gandro) [reply pkg/hubble/parser/sock/parser.go:130]: No, we did not. But this is a discussion that goes beyond this PR, so I'll keep this as is right now. 

Let's have the Hubble code handle this consistently in all packages. I don't want the socklb parser to do it differently than the other parsers do.

File: pkg/hubble/parser/sock/parser.go
Comment (gandro) [reply pkg/hubble/parser/sock/parser.go]: @aditighag PTAL. I've changed the code to use TRANSLATED for post events and the newly introduced TRACED verdict for pre events (as discussed with the Hubble team a week ago, we want a verdict here. Verdict UNKNOWN will be confusing to users, since it indicates missing data):

https://github.com/cilium/cilium/blob/c5372514225a0335457d65fb66fdec8b72cf03b3/pkg/hubble/parser/sock/parser.go#L215-L226

I've also added a short comment to each verdict value to document what it does.

**Review by qmonnet — APPROVED** [review]

**Review by aditighag — APPROVED** [review]

**Review by squeed — APPROVED** [review]
Looks good for my files!

**Review by rolinh — APPROVED** [review]
:shipit:

**Comment by gandro** [comment]
Removed `api` code owner, as it has been superseeded by `sig-hubble-api` and is thus not required anymore (if it was required, I could not have removed the review request).

Marking ready to merge.

**Pull request merged**
